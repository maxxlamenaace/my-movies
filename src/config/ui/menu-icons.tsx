import React, { ReactElement } from 'react';

import HomeOutlinedIcon from '@mui/icons-material/HomeOutlined';
import SlideshowOutlinedIcon from '@mui/icons-material/SlideshowOutlined';
import LiveTvOutlinedIcon from '@mui/icons-material/LiveTvOutlined';
import FavoriteBorderOutlinedIcon from '@mui/icons-material/FavoriteBorderOutlined';
import SearchOutlinedIcon from '@mui/icons-material/SearchOutlined';

export type MenuIcon = {
  display: 'home' | 'movies' | 'series' | 'search' | 'favorites';
  path: string;
  icon: ReactElement;
};

export const menuIcons: MenuIcon[] = [
  {
    display: 'home',
    path: '/',
    icon: <HomeOutlinedIcon />,
  },
  {
    display: 'movies',
    path: '/medias/movie',
    icon: <SlideshowOutlinedIcon />,
  },
  {
    display: 'series',
    path: '/medias/tv',
    icon: <LiveTvOutlinedIcon />,
  },
  {
    display: 'search',
    path: '/search',
    icon: <SearchOutlinedIcon />,
  },
  {
    display: 'favorites',
    path: '/favorites',
    icon: <FavoriteBorderOutlinedIcon />,
  },
];
