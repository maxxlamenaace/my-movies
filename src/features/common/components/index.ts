import Loader from './loader/Loader';
import Logo from './logo/Logo';
import Footer from './footer/Footer';
import AppBar from './app-bar/AppBar';
import Container from './container/Container';
import ScrollableAppBar from './scrollable-app-bar/ScrollableAppBar';
import Sidebar from './sidebar/Sidebar';
import ManualSwiper from './manual-swiper/ManualSwiper';
import Wrapper from './wrapper/Wrapper';
import PaginatorSwiper from './paginator-swiper/PaginatorSwiper';

export {
  Loader,
  Logo,
  Footer,
  AppBar,
  Container,
  ScrollableAppBar,
  Sidebar,
  ManualSwiper,
  Wrapper,
  PaginatorSwiper,
};
