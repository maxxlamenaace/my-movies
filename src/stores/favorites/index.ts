export * from './types';
export * from './favorites';
