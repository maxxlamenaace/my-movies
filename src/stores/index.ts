import { useAppStore } from './app/app';
import { useThemeModeStore } from './theme-mode/theme-mode';
import { useFavoritesStore } from './favorites/favorites';

export { useAppStore, useFavoritesStore, useThemeModeStore };
