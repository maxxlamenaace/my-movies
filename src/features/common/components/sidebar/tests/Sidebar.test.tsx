import React from 'react';

import { BrowserRouter } from 'react-router-dom';
import { cleanup, render, act, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Sidebar } from '@/features/common';
import { menuIcons } from '@/config/ui/menu-icons';
import { useThemeModeStore } from '@/stores';

describe('Sidebar', () => {
  const initialThemeStoreState = useThemeModeStore.getState();
  const toggleSidebarMock = jest.fn();

  beforeEach(() => {
    useThemeModeStore.setState(initialThemeStoreState, true);
  });

  afterEach(() => {
    cleanup();
  });

  it('renders correctly', () => {
    const { baseElement } = render(<Sidebar open={true} toggleSidebar={toggleSidebarMock} />, {
      wrapper: BrowserRouter,
    });

    expect(baseElement).toMatchSnapshot();
  });

  it('renders with as many buttons menus as needed', () => {
    const { baseElement } = render(<Sidebar open={true} toggleSidebar={toggleSidebarMock} />, {
      wrapper: BrowserRouter,
    });

    expect(baseElement.querySelectorAll('#sidebar-menu-button')).toHaveLength(menuIcons.length);
  });

  it('renders with a logo', () => {
    const { baseElement } = render(<Sidebar open={true} toggleSidebar={toggleSidebarMock} />, {
      wrapper: BrowserRouter,
    });

    expect(baseElement.querySelector('#logo')).toBeDefined();
  });

  describe('when user clicks on switch theme mode', () => {
    it('updates theme mode', async () => {
      const { baseElement } = render(<Sidebar open={true} toggleSidebar={toggleSidebarMock} />, {
        wrapper: BrowserRouter,
      });

      const swithThemeButton = baseElement.querySelector('#sidebar-switch-theme');

      expect(useThemeModeStore.getState().mode).toEqual('dark');
      expect(swithThemeButton).toBeDefined();

      act(() => {
        if (swithThemeButton) {
          fireEvent.click(swithThemeButton);
        }
      });

      await waitFor(() => expect(useThemeModeStore.getState().mode).toEqual('light'));
    });
  });

  describe('when user triggers toggle sidebar', () => {
    it('calls the toggleSidebar function', () => {
      render(<Sidebar open={true} toggleSidebar={toggleSidebarMock} />, {
        wrapper: BrowserRouter,
      });

      act(() => {
        userEvent.keyboard('{esc}');
      });

      expect(toggleSidebarMock).toHaveBeenCalledTimes(1);
    });
  });
});
