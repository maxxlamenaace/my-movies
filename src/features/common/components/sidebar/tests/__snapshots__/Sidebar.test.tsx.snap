// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders correctly 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDrawer-root MuiDrawer-modal MuiModal-root css-1ys78cs-MuiModal-root-MuiDrawer-root"
    id="sidebar-container"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation16 MuiDrawer-paper MuiDrawer-paperAnchorLeft css-4t3x6l-MuiPaper-root-MuiDrawer-paper"
      tabindex="-1"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-14hsq2w-MuiToolbar-root"
      >
        <div
          class="css-2hk24g-MuiStack-root"
        >
          <a
            href="/"
            id="logo-link"
            style="text-decoration: none;"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-10bvdjj-MuiTypography-root"
              id="logo"
              style="color: rgba(0, 0, 0, 0.87);"
            >
              my
              <span
                style="color: rgb(25, 118, 210);"
              >
                Movies
              </span>
            </p>
          </a>
        </div>
      </div>
      <ul
        class="MuiList-root MuiList-padding css-c6jbv8-MuiList-root"
      >
        <a
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-khacv1-MuiButtonBase-root-MuiListItemButton-root"
          href="/"
          id="sidebar-menu-button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-whim71-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="HomeOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m12 5.69 5 4.5V18h-2v-6H9v6H7v-7.81l5-4.5M12 3 2 12h3v8h6v-6h2v6h6v-8h3L12 3z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-l1zdit-MuiTypography-root"
            >
              home
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
        <a
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-ewosjp-MuiButtonBase-root-MuiListItemButton-root"
          href="/medias/movie"
          id="sidebar-menu-button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-qkke2n-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SlideshowOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 8v8l5-4-5-4zm9-5H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V5h14v14z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1oia209-MuiTypography-root"
            >
              movies
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
        <a
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-ewosjp-MuiButtonBase-root-MuiListItemButton-root"
          href="/medias/tv"
          id="sidebar-menu-button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-qkke2n-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="LiveTvOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9 10v8l7-4zm12-4h-7.58l3.29-3.29L16 2l-4 4h-.03l-4-4-.69.71L10.56 6H3c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h18c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2zm0 14H3V8h18v12z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1oia209-MuiTypography-root"
            >
              series
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
        <a
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-ewosjp-MuiButtonBase-root-MuiListItemButton-root"
          href="/search"
          id="sidebar-menu-button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-qkke2n-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1oia209-MuiTypography-root"
            >
              search
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
        <a
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-ewosjp-MuiButtonBase-root-MuiListItemButton-root"
          href="/favorites"
          id="sidebar-menu-button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-qkke2n-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FavoriteBorderOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1oia209-MuiTypography-root"
            >
              favorites
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
      </ul>
      <div
        class="MuiBox-root css-16dnimb"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-w5a47t-MuiButtonBase-root-MuiIconButton-root"
          id="sidebar-switch-theme"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DarkModeOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M9.37 5.51c-.18.64-.27 1.31-.27 1.99 0 4.08 3.32 7.4 7.4 7.4.68 0 1.35-.09 1.99-.27C17.45 17.19 14.93 19 12 19c-3.86 0-7-3.14-7-7 0-2.93 1.81-5.45 4.37-6.49zM12 3c-4.97 0-9 4.03-9 9s4.03 9 9 9 9-4.03 9-9c0-.46-.04-.92-.1-1.36-.98 1.37-2.58 2.26-4.4 2.26-2.98 0-5.4-2.42-5.4-5.4 0-1.81.89-3.42 2.26-4.4-.44-.06-.9-.1-1.36-.1z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
